### PR TITLE
exit if the user has requested a specific port

### PR DESCRIPTION
**Pull request by SvenDowideit**
Signed-off-by: Sven Dowideit <sven.dowideit@csiro.au>

for https://github.com/datproject/dat/issues/945

### CODE REVIEW

**Comment by SvenDowideit** [comment]
mmm, ok, no, this won't work as is, looking deeper.

**Comment by SvenDowideit** [comment]
actually, playing around more, no changes may be needed in dat-node at all >yay newbies<

File: lib/network.js [hunk 1]
```diff
@@ -13,7 +13,9 @@ module.exports = function (archive, opts, cb) {
     stream: opts.stream
   }, opts)
   var swarm = disc(swarmDefaults(swarmOpts))
-  swarm.once('error', function () {
+  swarm.once('error', function (err) {
+    // Don't magically use any port if the users has requested a specific port
+    swarm.emit('errorport', err)
```
Comment (joehand) [inline lib/network.js]: Can also we call the callback, `cb`?

File: lib/network.js [hunk 2]
```diff
@@ -13,7 +13,9 @@ module.exports = function (archive, opts, cb) {
...
     swarm.listen(0)
```
Comment (joehand) [inline lib/network.js]: Should we call this if `opts.port` isn't set?

**Comment by joehand** [comment]
It'd be great to have a test for this.

File: lib/network.js (hunk 2 above)
Comment (SvenDowideit) [reply lib/network.js]: I originally added an if for that too - but the errorport handling shortcircuits before that, so it seemed cleaner to not add a duplicate if

File: lib/network.js (hunk 1 above)
Comment (SvenDowideit) [reply lib/network.js]: mmm, looking at that file - no idea, I don't know what it does, or how / when it should be called :/

Comment (SvenDowideit) [reply lib/network.js]: ah, ok, that changes the if answer too - awesome, updating.

Comment (joehand) [reply lib/network.js]: > mmm, looking at that file - no idea, I don't know what it does, or how / when it should be called :/

The callback can be used to know if there are any network connections (its called back after first round of discovery). But we can pass the error back as the first argument as well.

File: lib/network.js [hunk 3]
```diff
@@ -13,7 +13,13 @@ module.exports = function (archive, opts, cb) {
     stream: opts.stream
   }, opts)
   var swarm = disc(swarmDefaults(swarmOpts))
-  swarm.once('error', function () {
+  swarm.once('error', function (err) {
+    // Don't magically use any port if the users has requested a specific port
+    swarm.emit('errorport', err)
```
Comment (joehand) [inline lib/network.js]: Can we use `emit('error', new Error('specified port not available'))` (or similar) for use cases outside the dat cli. Will require less documentation than a custom event =).

(Not sure if I left this before but not seeing it. Apologies if its duplicate.)

**Comment by joehand** [comment]
I updated the CLI PR for this to avoid a custom error message. It'd still be good to have this in dat-node but will leave it to user for now.

**Pull request closed without merging**